vcd/resource_vcd_vapp_access_control.go

```diff
@@ -0,0 +1,360 @@
+		Update: resourceAccessControlVappUpdate,
+		Delete: resourceAccessControlVappDelete,
+		Importer: &schema.ResourceImporter{
+			State: accessControlVappImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"org": {
+				Type:     schema.TypeString,
+				Required: false,
+				Optional: true,
+				ForceNew: true,
+				Description: "The name of organization to use, optional if defined at provider " +
+					"level. Useful when connected as sysadmin working across different organizations",
+			},
+			"vdc": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			"vapp_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "vApp identifier",
+			},
+			"shared_to_everyone": &schema.Schema{
+				Type:        schema.TypeBool,
+				Required:    true,
+				Description: "Whether the vApp is shared to everyone",
+			},
+			"everyone_access_level": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{types.ControlAccessReadOnly, types.ControlAccessReadWrite, types.ControlAccessFullControl}, true),
+				Description:  "Access level when the vApp is shared with everyone (one of ReadOnly, Change, FullControl). Required when shared_to_everyone is set",
+			},
+			"shared": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				MinItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"user_id": &schema.Schema{
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "ID of the user to which we are sharing. Required if group_id is not set",
+						},
+						"group_id": &schema.Schema{
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "ID of the group to which we are sharing. Required if user_id is not set",
+						},
+						"subject_name": &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Name of the subject (group or user) with which we are sharing",
+						},
+						"access_level": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringInSlice([]string{types.ControlAccessReadOnly, types.ControlAccessReadWrite, types.ControlAccessFullControl}, true),
+							Description:  "The access level for the user or group to which we are sharing. (One of ReadOnly, Change, FullControl)",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// tenantContext defines whether we run access control operations in the context of the original caller.
+// By default it is ON (= run as tenant). We can turn it off by setting the environment variable VCD_ORIGINAL_CONTEXT.
+var tenantContext = os.Getenv("VCD_ORIGINAL_CONTEXT") == ""
```

Comment:
I'm still confused:
Assume I have a user in the system ORG. 
Assume I update a ACL in ORG1234 with this user.
Then tenantContext is the ORG1234?

Then if I set env `VCD_ORIGINAL_CONTEXT=true` then I would assume it runs in system context? So `tenantContext = false` ? 

Am I wrong?

Reply:
When tenant context is ON (the default) we add these two items to the request header.
```
X-VMWARE-VCLOUD-TENANT-CONTEXT: ${Org-ID}
X-VMWARE-VCLOUD-AUTH-CONTEXT: ${Org-Name}
```
The context is the org to which the vApp belongs.

When `VCD_ORIGINAL_CONTEXT` is set, the headers are not added.

Reply:
Ah, got it now. Any arbitrary value in `VCD_ORIGINAL_CONTEXT` will work to remove tenant context.
OK
LGTM again